crates/sui-rpc-api/proto/sui/rpc/v2alpha/move_descriptor.proto

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+syntax = "proto3";
+
+package sui.rpc.v2alpha;
+
+// Describes a data type.
+message Descriptor {
+  // Name of this DataType
+  optional string name = 1;
+
+  // XXX Do we want both storage/runtime info here??
+  // PackageId where this DataType is defined.
+  optional string package_id = 2;
+
+  // Name of the module where this DataType is defined
+  optional string module = 3;
+
+  // Fully qualified name of this DataType.
+  //
+  // This is `<package_id>::<module>::<name>`
+  optional string type_name = 4;
+
+  // This type's abilities
+  repeated Ability abilities = 4;
+
+  // Ability constraints and phantom status for this type's generic type parameters
+  repeated TypeParameter type_parameters = 5;
+
+  // XXX do we use oneof to differentiate or an enum?
+  // We'd have to use a wrapper message. Or we have a separate message type for enumes/structs
+
+  // Set of fields if this DataType is a struct.
+  //
+  // The order of the entries is the order of how the fields are defined.
+  repeated FieldDescriptor fields = 10;
+
+  // Set of variants if this DataType is an enum.
+  //
+  // The order of the entries is the order of how the variants are defined.
+  repeated VariantDescriptor variants = 11;
+}
+
+// An `Ability` classifies what operations are permitted for a given type
+enum Ability {
+  ABILITY_UNKNOWN = 0;
+
+  // Allows values of types with this ability to be copied
+  COPY = 1;
+
+  // Allows values of types with this ability to be dropped.
+  DROP = 2;
+
+  // Allows values of types with this ability to exist inside a struct in global storage
+  STORE = 3;
+
+  // Allows the type to serve as a key for global storage operations
+  KEY = 4;
+}
+
+// A generic type parameter used in the declaration of a struct or enum.
+message TypeParameter {
+  // The type parameter constraints
+  repeated Ability constraints = 1;
+
+  // Whether the parameter is declared as phantom
+  optional bool is_phantom = 2;
+}
+
+// Descriptor of a field that belongs to a struct or enum variant
+message FieldDescriptor {
+  // Name of the field
+  optional string name = 1;
+
+  // Order or position of the field in the struct or enum variant definition.
+  optional uint32 position = 2;
+
+  // The type of the field
+  optional OpenSignatureBody type = 5;
+}
+
+enum Type {
+  TYPE_UNKNOWN = 0;
+  ADDRESS = 1;
+  BOOL = 2;
+  U8 = 3;
+  U16 = 4;
+  U32 = 5;
+  U64 = 6;
+  U128 = 7;
+  U256 = 8;
+  VECTOR = 9;
+  DATA_TYPE = 10;
```

Comment:
nit (requires clarification): I vaguely recall @tzakian telling me that "datatype" was one word in this case, but I don't remember the details beyond that.

Reply:
Yea, it should all be one word in my view -- in code/pretty much everywhere we use `Datatype` or `datatype`, which would correspond to `DATATYPE` here. So +1 to one word.